cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/LineageAdmin.java

```diff
@@ -251,35 +167,133 @@ private Lineage doComputeLineage(final NamespacedEntityId sourceData,
+        appSpecs.computeIfAbsent(relation.getProgram().getParent(), store::getApplication);
+      }
+    });
+
+    // Step 2, get the run record for all the possible inner programs, the run record contains the
+    // workflow information, fetch the workflow id and add them to the map
+    Map<ProgramRunId, RunRecordMeta> runRecords = store.getRuns(possibleInnerPrograms);
+    Set<ProgramRunId> workflowRunIds = new HashSet<>();
+    runRecords.entrySet().stream()
+      .filter(e -> e.getValue() != null)
+      .forEach(entry -> {
+        ProgramRunId programRunId = entry.getKey();
+        RunRecordMeta runRecord = entry.getValue();
+
+        if (runRecord.getSystemArgs().containsKey(ProgramOptionConstants.WORKFLOW_RUN_ID)) {
+          ProgramRunId wfRunId = extractWorkflowRunId(programRunId, runRecord);
+          programWorkflowMap.put(programRunId, wfRunId);
+          workflowRunIds.add(wfRunId);
+        }
+      }
+    );
+
+    // Step 3, fetch run records of the workflow, the properties of the workflow run record has all
+    // the inner program run ids, compare them with the app spec to get the type of the program
+    runRecords = store.getRuns(workflowRunIds);
+    runRecords.entrySet().stream()
+      .filter(e -> e.getValue() != null)
+      .forEach(entry -> {
+        ProgramRunId programRunId = entry.getKey();
+        RunRecordMeta runRecord = entry.getValue();
+        extractAndAddInnerPrograms(toVisitPrograms, programWorkflowMap, appSpecs, programRunId, runRecord);
+      });
+  }
+
+  private ProgramRunId extractWorkflowRunId(ProgramRunId programRunId, RunRecordMeta runRecord) {
+    Map<String, String> systemArgs = runRecord.getSystemArgs();
+    String workflowRunId = systemArgs.get(ProgramOptionConstants.WORKFLOW_RUN_ID);
+    String workflowName = systemArgs.get(ProgramOptionConstants.WORKFLOW_NAME);
+    WorkflowId workflowId = programRunId.getParent().getParent().workflow(workflowName);
+    return workflowId.run(workflowRunId);
+  }
+
+  /**
+   * Extract inner programs and runs from the workflow run record, the run record's properties have all the
+   * inner program run ids. The workflow spec can then be used to determine what the inner programs are and
+   * create the program run ids for them
+   */
+  private void extractAndAddInnerPrograms(Set<ProgramId> toVisitPrograms,
+                                          Map<ProgramRunId, ProgramRunId> programWorkflowMap,
+                                          Map<ApplicationId, ApplicationSpecification> appSpecs,
+                                          ProgramRunId programRunId, RunRecordMeta wfRunRecord) {
+    wfRunRecord.getProperties().forEach((key, value) -> {
+      ApplicationId appId = programRunId.getParent().getParent();
+      WorkflowSpecification workflowSpec =
+        appSpecs.get(appId).getWorkflows().get(programRunId.getProgram());
+      Map<String, WorkflowNode> nodeIdMap = workflowSpec.getNodeIdMap();
```

Comment:
These three variables should be outside the `forEach`, as they don't depends on `(key, value)`.